clippy_lints/src/returns.rs

```diff
@@ -268,8 +284,9 @@ fn check_final_expr<'tcx>(
         // note, if without else is going to be a type checking error anyways
         // (except for unit type functions) so we don't match it
         ExprKind::Match(_, arms, MatchSource::Normal) => {
+            let match_ty = cx.typeck_results().expr_ty(peeled_drop_expr);
             for arm in arms.iter() {
-                check_final_expr(cx, arm.body, semi_spans.clone(), RetReplacement::Unit);
+                check_final_expr(cx, arm.body, semi_spans.clone(), RetReplacement::Empty, Some(match_ty));
```

Comment:
I might be mistaken, but I feel like this is not quite the right fix for the issue. The problem, as I understand it, is that the `return` value has a different type than the other match arms. From my understanding, this can only happen, if the match is at the end of the function and has a semicolon, making it a statement. This therefore also implies that the function returns the unit type `()`, as otherwise there would be something after the statement or the types would be compatible.

```rs
fn test_match_as_stmt() {
    let x = 9;
    match x {
        1 => 2,
        _ => return,
    }; // The semicolon makes it a statement, and allows the branches to be non-unit
    // This makes the branches incompatible
}

fn does_not_compile() {
    let x = 9;
    match x {
        1 => 2,
        _ => return,
    }
    // The first branch is incompatible with the function signature.
}

fn does_compile_because_types_are_compatible() -> u32 {
    let x = 9;
    match x {
        1 => 2,
        _ => return 3,
    }
    // Should be linted since the types are compatible
}
```

(Maybe there are some other edge cases with `impl Trait`, that might be worth testing)

Based on this, I think the proper fix would be to check the type of the match statement. If it's `()` the types are compatible. If not and the `return` has a value than they would still be compatible. It would only cause problems, if the `match` type is `!= ()` while the function is `-> ()`.

Does this make sense?

Reply:
> I think the proper fix would be to check the type of the match statement
 
 That's what I am doing currently in the function `check_final_expr` here:
 https://github.com/feniljain/rust-clippy/blob/c12748fab3a14beae4958f13551e7e6c52298490/clippy_lints/src/returns.rs#L250-L260
 
 😅 
 
 > If not and the return has a value than they would still be compatible.
 
 We would not reach this ( one I linked just above ) if the return is not empty, in which case we just tell clippy to ignore it, at least that's what we discussed in zulip